squeal-postgresql-ltree/src/Squeal/PostgreSQL/LTree.hs

```diff
@@ -0,0 +1,412 @@
+%           Match initial underscore-separated words
+@
+
+The behavior of % is a bit complicated.
+It tries to match words rather than the entire label.
+For example foo_bar% matches foo_bar_baz but not foo_barbaz.
+If combined with *, prefix matching applies to each word separately,
+for example foo_bar%* matches foo1_bar2_baz but not foo1_br2_baz.
+
+Also, you can write several possibly-modified labels separated with
+| (OR) to match any of those labels,
+and you can put ! (NOT) at the start to match any label
+that doesn't match any of the alternatives.
+
+Here's an annotated example of lquery:
+
+@
+Top.*{0,2}.sport*@.!football|tennis.Russ*|Spain
+1.  2.     3.      4.               5.
+@
+
+This query will match any label path that:
+
+1. begins with the label Top
+2. and next has zero to two labels before
+3. a label beginning with the case-insensitive prefix sport
+4. then a label not matching football nor tennis
+5. and then ends with a label beginning with Russ or exactly matching Spain.
+-}
+newtype LQuery = UnsafeLQuery {getLQuery :: Text}
+  deriving stock (Eq,Ord,Show,Read,Generic)
+-- | `PGlquery`
+instance IsPG LQuery where type PG LQuery = PGlquery
+instance FromPG LQuery where
+  fromPG = UnsafeLQuery <$> devalue Decoding.text_strict
+instance ToPG db LQuery where
+  toPG = pure . Encoding.text_strict . getLQuery
+instance Inline LQuery where
+  inline
+    = UnsafeExpression
+    . parenthesized
+    . (<> " :: lquery")
+    . escapeQuotedText
+    . getLQuery
+
+{- |
+ltxtquery represents a full-text-search-like pattern for matching ltree values.
+An ltxtquery value contains words,
+possibly with the modifiers @, *, % at the end;
+the modifiers have the same meanings as in lquery.
+Words can be combined with & (AND), | (OR), ! (NOT), and parentheses.
+The key difference from lquery is that ltxtquery matches words
+without regard to their position in the label path.
+
+Here's an example ltxtquery:
+
+@
+Europe & Russia*@ & !Transportation
+@
+
+This will match paths that contain the label Europe and any label
+beginning with Russia (case-insensitive), but not paths containing
+the label Transportation. The location of these words within the
+path is not important. Also, when % is used, the word can be matched
+to any underscore-separated word within a label, regardless of position.
+
+Note: ltxtquery allows whitespace between symbols, but ltree and lquery do not.
+-}
+newtype LTxtQuery = UnsafeLTxtQuery {getLTxtQuery :: Text}
+  deriving stock (Eq,Ord,Show,Read,Generic)
+-- | `PGltxtquery`
+instance IsPG LTxtQuery where type PG LTxtQuery = PGltxtquery
+instance FromPG LTxtQuery where
+  fromPG = UnsafeLTxtQuery <$> devalue Decoding.text_strict
+instance ToPG db LTxtQuery where
+  toPG = pure . Encoding.text_strict . getLTxtQuery
+instance Inline LTxtQuery where
+  inline
+    = UnsafeExpression
+    . parenthesized
+    . (<> " :: ltxtquery")
+    . escapeQuotedText
+    . getLTxtQuery
+
+instance IsString
```

Comment:
hm, these instances are convenient but also a dangerous since they allow for constructing invalid values.

Reply:
Yes, I think this is ok. Squeal doesn't guarantee no exceptions.